function.k

```diff
@@ -0,0 +1,249 @@
+         <context> CRho:Map </context>
+         ...
+     </account>
+
+
+rule [Internal-Function-Call]:
+     <k> functionCall(F:Id; Vs:Values) => #Call(F,Vs) ~> #return ... </k>
+     <env> ENV:Map  </env>
+     <localCalls> .List => ListItem(#state(ENV,#return(false,#undef_Value))) ... </localCalls> 
+
+rule 
+     <k> #Call(F:Id, Vs:Values) => #BindInputParams(Ps, Vs) ~> #BindReturnParams(OPs, .Values) ~> #CallBody(Ss) ...</k>
+     <currentAccount> #account(C:Int,N:Id) </currentAccount>
+     <Contract>
+         <cname> N:Id </cname>
+         <Function>
+             <fname> F </fname>
+             <inputParams> Ps:FParameters </inputParams>
+             <Returns> OPs:FParameters </Returns>
+             <body> Ss:Statements </body>
+             ...
+         </Function>
+         ...
+     </Contract>
+
+rule
+     <k> #BindInputParams(Ps, Vs) => #InitParams(Ps, Vs) ...</k>
+
+rule
+     <k> #BindReturnParams(OPs, Vs) => #InitParams(OPs, Vs) ...</k>
+
+     
+rule 
+     <k> #InitParams(P:FParameter, Ps:FParameters, V:Value,Vs:Values)
+     	    => #InitParam(P, V) ~> #InitParams(Ps, Vs) ...</k>
+
+rule 
+     <k> #InitParams(P:FParameter, Ps:FParameters, .Values)
+     	    => #InitParam(P, #undef_Value) ~> #InitParams(Ps, .Values) ...</k>
+
+rule 
+     <k> #InitParams(.FParameters, .Values) => . ...</k>
+
+rule 
+     <k> #InitParam(T:TypeName S:StorageLocations X:Id, V) => #allocate(C:Int, CN, #varInfo(X:Id, V:Value, T:TypeName, #mem)) ...</k>
+     <currentAccount> #account(C:Int,CN:Id) </currentAccount>
+
+rule 
+     <k> #InitParam(T:TypeName, V) => . ...</k>
+
+rule [Call-Function-Body]:
+     <k> #CallBody(S:Statement Ss:Statements) => #exeStmt(S) ~> #CallBody(Ss) ...</k>
+
+rule 
+     <k> #CallBody(.Statements) => . ...</k>
+
+rule
+    <k> #exeStmt(S:SimpleStatement;) => S; ...</k>
+    <localCalls> ListItem(#state(_:Map,#return(false,_:Value))) ... </localCalls>
+
+rule
+     <k> #exeStmt(S:SimpleStatement;) => . ...</k>
+     <localCalls> ListItem(#state(_:Map,#return(true,_:Value))) ... </localCalls>
+
+rule
+     <k> #exeStmt(S:SimpleStatement;) => S; ...</k>
+     <localCalls> .List </localCalls>
+     <callStack> ListItem(#state(#account(_:Int,_:Id), _:Map, #return(false,_:Value))) ... </callStack>
+
+rule
+     <k> #exeStmt(S:SimpleStatement;) => . ...</k>
+     <localCalls> .List </localCalls>
+     <callStack> ListItem(#state(#account(_:Int,_:Id), _:Map, #return(true,_:Value))) ... </callStack>
+
+rule [Local-Call-Return]:
+     <k> #return  => V  ... </k>
+     <localCalls> ListItem(#state(Rho,#return(_:Bool,V:Value))) => .List ... </localCalls> 
+     <env> _ => Rho </env>
+
+rule [External-Call-Return]:
+     <k> #returnContract => V ... </k>
+     <callStack> ListItem(#state(C, Rho, #return(_:Bool,V:Value))) => .List ... </callStack>
+     <env> _ => Rho </env>
+     <currentAccount> _ => C  </currentAccount>
+     <callDepth> Depth => Depth -Int 1 </callDepth>
+
+rule [ReturnValue-In-ExternalCall]:
```

Comment:
https://github.com/kframework/solidity-semantics/blob/8b967ad6b17ae8413fe204a69f0e7cb9bfa61aa4/function.k#L231-L233
https://github.com/kframework/solidity-semantics/blob/8b967ad6b17ae8413fe204a69f0e7cb9bfa61aa4/function.k#L240-L242
I think these rules are wrong because they can be applied even when the `return` is for a external call.
Try this
```solidity
contract A {
        int a;
	function g(int aa) public returns (int) {
          a = aa + 1;
          return a;
	}
}

contract B {
        int b;
	function g1() public returns (int) {
          functionCall(g2;);
          return b;
	}
        function g2() internal {
          b = functionCall(new A(); g; 123) ; // this function returns nothing
        }
}

{
"Test1" : {
	"exec" : {
	    "code" : #solidity( int ans; ans = functionCall(new B() ; g1 ; ); )
	},
        "post" : {
            "mem" : #exists("ans", 124)

        }
     }
}
```
If you krun this with `--search-all` flag, you can find
```
<localCalls>
  ListItem ( #state ( b |-> #storedVar ( 2 , int , #storage ) , #return ( true , 124 ) ) )
</localCalls>
```
and this doesn't make sense.

Reply:
Actually, this is not an intended behavior of external function calls.  Is it possible that external calls can be nested in local calls? Instead of calling g1, you can directly call g2. These rules are based on the assumptions that external calls cannot be nested in local calls.

Reply:
Btw,  functionCall(new A(); g; 123) ; and functionCall(g2;); are our internal interpretations and not valid syntax in Solidity.

Reply:
```solidity
contract A {
    int a;
    function g(int aa) public returns (int) {
    	a = aa + 1;
      	return a;
    }
}

contract B {
    int b;
    function g1() public returns (int) {
   	g2(); 
      	return b;
    }
    function g2() internal {
      	b = (new A()).g(123);
    }
}


```
this is a valid solidity program.

Reply:
If this is possible, then how can we know whether the most inside call is a local call or external call?

Reply:
Hmm. Looks like the idea behind those rules is "if `<localCalls>` is empty then we are in an external call; otherwise we are in an internal call". But unfortunately it seems not to be that way, i.e. we can be in an external call _and_ have `<localCall>` not empty. In that case, rule `[ReturnValue-In-ExternalCall]` can't apply (because it requires an empty `<localCall>`) and rule `[ReturnValue-In-LocalCall]` applies instead, changing the content of `<localCalls>` in an unexpected way. 

I guess the question then is: how to properly distinguish between the external vs internal call cases?

Reply:
Btw: @tomtomjhj I tried running the example with `--search-all` as you suggested, but I noticed the output is ~36K lines long. I wonder, how did you find out about that? Did you manually inspect all the output or did you search for something?

Reply:
The previous assumption is that locals can be nested in external calls, but external calls cannot be nested in local calls. However, they can be present at the same time. If we use one stack for all function calls, the problem will be solved.

Reply:
possible solutions
* just don't use `<localCalls>`. Put everything in `<callStack>` and put some makers on the elements
* 
```
rule [External-Function-Call]:
      ......
     <localCalls> LocalCallStack => .List </localCalls>
     <callStack> .List => ListItem(#state(#account(C1,N1), Rho, LocalCallStack, #return(false,#undef_Value))) ... </callStack>
     ......

rule [Return-In-ExternalCall]:
     <k> return  => #end_Exp ... </k>
     <localCalls> .List => LocalCallStack </localCalls>
     <callStack> ListItem(#state(C, Rho, LocalCallStack #return(_:Bool => true, _:Value))) ... </callStack>
......
```
with these modifications, a `return` is for internal func iff `<localCalls>` is not empty.

Reply:
@dfilaretti I first suspected that there might be some issues with this, wrote/ran the example code, and just used vim to find that line.

Reply:
@JIAOJIAO1991 agreed.

Reply:
So which approach have you decided to take? 

At a first look, both possible solutions proposed by by @tomtomjhj look reasonable to me. 
Maybe solution (2) will require less changes to the current code (since you get to keep both stacks)? But at the same time solution (1) requires 1 less cell and is probably simpler (you just add some tag to the stack frame and look at that to decide whether it's internal or external). 

I'm not sure what's better though. 
It would be great to hear someone else's thoughts on this. @dwightguth @ehildenb do you have any advice to share? Thank you!

Reply:
I suggest option one, i.e., using only one stack `callStack` with addition informative markers to distinguish between external or internal function calls.

The reason is that a stack is supposed to help us to keep track of the thing happens most recently (last in first out). If we use two stacks, then we lose the order relation of elements between two stacks, that is, which one happens the most recently, the top of stack 1 or the top of stack 2? Of course we can put some addition effort to maintain the order/sequence, but it somehow complicates the flow.

What do you think? @dwightguth, @ehildenb, and @tomtomjhj.

Reply:
@shangweilin what about an extra cell inside `callState` marking if it's a local or external call? Similar to how the `static` cell works, it will just be a single `Bool` that you need to check.

Then, when you do `#pushCallStack`, because the entire contents of `callState` will be saved, it will also save off that particular value.

Reply:
We have fixed this issue by removing the cell `localCalls`. That is, we only have one stack now, which is `<callStack>`. In addition, we add `FunctionFlag`, which could be `#internal` or `#external`, into each stack frame, so that we can distinguish between external/local function calls.

Reply:
Marking as resolved to keep conversations concise. I have additional comments below about this subject that I think should be resolved.